index.bs

```diff
@@ -4109,165 +4310,1257 @@ To <dfn>get the response data</dfn> given |response|:
+
+    1. If |pattern| [=map/contains=] "<code>search</code>":
+
+      1. Let |search| be the result of [=trying=] to [=unescape URL pattern=] with
+         |pattern|["<code>search</code>"].
+
+      1. If |search| does not [=string/starts with|start with=] U+003F (?), then
+         append "<code>?</code>" to |pattern url|.
+
+      1. For each |codepoint| in |search|:
+
+        1. If |codepoint| is U+0023 (#):
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Append |search| to |pattern url|.
+
+1. Otherwise, if |pattern| matches the <code>network.UrlPatternString</code> production:
+
+  1. Let |pattern url| be the result of [=trying=] to [=unescape URL
+     pattern=] with |pattern|["<code>pattern</code>"].
+
+1. Let |url| be the result of [=URL Parser|parsing=] |pattern url|.
+
+1. If |url| is failure, return [=error=] with [=error code=] [=invalid argument=].
+
+1. Let |parsed| be a struct with the following fields:
+   <dl>
+     <dt>protocol
+     <dd>|url|'s [=url/scheme=] if |has protocol| is true, or null otherwise.
+     <dt>hostname
+     <dd>|url|'s [=url/host=] if |has hostname| is true, or null otherwise.
+     <dt>port
+     <dd>
+     1. If |url|'s [=url/scheme=] [=is special=] and |url|'s
+   [=url/scheme=]'s [=default port=] is not null, and |url|'s [=url/port=] is
+   null or is equal to [=url/scheme=]'s [=default port=]:
+
+       1. The empty string.
+
+     1. Otherwise, if |url|'s [=url/port=], is not null:
+
+       1. [=Serialize an integer=] with |url|'s [=url/port=].
+
+     1. Otherwise:
+
+       1. null.
+
+     <dt>pathname
+     <dd>The result of running the [=URL path serializer=] with |url|, if
+         |url|'s [=url/path=] is not the empty string and is not [=list/empty=],
+         or null otherwise.
+     <dt>search
+     <dd>|url|'s [=url/query=].
+   </dl>
+
+1. Return [=success=] with data |parsed|.
+
+</div>
+
+<div algorithm>
+To <dfn>match URL pattern</dfn> given |url pattern| and |url string|:
+
+1. Let |url| be the result of [=URL parser|parsing=] |url string|.
+
+1. If |url pattern|'s protocol is not null and is not equal to |url|'s
+   [=url/scheme=], return false.
+
+1. If |url pattern|'s hostname is not null and is not equal to |url|'s
+   [=url/host=], return false.
+
+1. If |url pattern|'s port is not null:
+
+   1. Let |port| be null.
+
+   1. If |url|'s [=url/scheme=] [=is special=] and |url|'s
+      [=url/scheme=]'s [=default port=] is not null, and |url|'s [=url/port=] is
+      null or is equal to [=url/scheme=]'s [=default port=]:
+
+     1. Set |port| to the emtpy string.
+
+   1. Otherwise, if |url|'s [=url/port=], is not null:
+
+     1. Set |port| to [=serialize an integer=] with |url|'s [=url/port=].
+
+   1. If |url pattern|'s port is not equal to |port|, return false.
+
+1. If |url pattern|'s pathname is not null and is not equal to the result of
+   running the [=URL path serializer=] with |url|, return false.
```

Comment:
Another interesting scenario, for a URLPatternPattern such as `{ protocol: "http" }`, since this is a special scheme, the basic URL parser will set the path to a single segment empty string `[""]`, meaning we will have a non-null pathname set to `/`. Such a URLPatternPattern will therefore match any URL using the `http` protocol and with an empty path. Meaning `http://abc.com:1234/?search` will match, but `http://abc.com:1234/path?search` will not. 

I feel like it's unexpected, and we probably want a `has path` logic here?